challenges-7-plus.yml

```diff
@@ -0,0 +1,277 @@
+        <h1>Welcome our Pizza Section!</h1>
+    
+        <p>You need to be a pizza lover before you proceed!</p>
+    
+        <button
+            class="btn btn-primary"
+            onclick="javascript:document.getElementById('my_great_modal').style.display='none';"
+        >
+            Proceed
+        </button>
+    </div>
+    ```
+    
+    I then use JavaScript to discretely show this popup under certain circumstances (i.e. when
+    I detect that a customer might not love pizza!).
+    
+    But, when the user leaves the page and then goes back, Turbo is showing this modal
+    *even* when the user might have proven otherwise that he is indeed a Pizzal lover!
+    I don't want that!
+    
+    How can I hide this modal every time Turbo shows my page preview?
+  explanation: |
+    Leveraging the turbo event `turbo:before-cache` is the best solution for things like
+    these, because you can clean up and prepare your snapshow before turbo takes it so that
+    you ensure that you will always see the right state!
+    
+    Other seemingly useful events, like `turbo:load` might look useful, but they will
+    only fire after the page has finished loading, which means that the preview event
+    already took place!
+  features:
+    multiple_choice:
+      choices:
+        a: |
+          ```javascript
+          document.addEventListener('turbo:before-visit', (event) => {
+              document.getElementById('my_great_modal').display = 'none';
+          });
+          ```
+        b: |
+          ```javascript
+          document.addEventListener('turbo:before-cache', (event) => {
+              document.getElementById('my_great_modal').display = 'none';
+          });
+          ```
+        c: |
+          ```javascript
+          document.addEventListener('turbo:before-fetch-request', (event) => {
+              document.getElementById('my_great_modal').display = 'none';
+          });
+          ```
+        d: |
+          ```javascript
+          document.addEventListener('turbo:load', (event) => {
+              document.getElementById('my_great_modal').display = 'none';
+          });
+          ```
+      correct: b
+
+cleaned_preview:
+  question: |
+    In our video, we show how closing a bootstrap dialog before exiting a page,
+    so that we can have a cleaned preview, doesn't quite work. When we came
+    back to the page, we saw a greyed invalidation div covering it!
+    
+    Fixing this required some more hacking on our part in order to have it working
+    properly.
+    
+    Why did we get this weird behavior?
+  explanation: |
+    The `turbo:before-cache` event is synchronous, which means that what you do in that
+    event handler will happen instantly. If in that event you call some animation
+    (like the bootstrap `close` function), then the animation won't finish executing,
+    preventing any callbacks after the animation from cleaning up the page.
```

Comment:
This might require a bit more explanation, since it's so subtle.

> , then the animation won't finish executing. This can cause two potential problems. First,
> the snapshot will be taken before the element that is animating (e.g. the modal closing) 
> is finished animating. And second, as in this case, Bootstrap is "listening" for the animation
> to finish so that it can do other clean up, like removing the overlay. But because the element
> was removed from the page, the animation never "finishes" and the cleanup never happens.
> Weird, eh?